### PR TITLE
Add EI support to TFS container.

### PR DESCRIPTION
**Pull request by ChuyangDeng**
*Issue #, if available:*
https://github.com/awslabs/amazon-sagemaker-examples/issues/613
https://github.com/aws/sagemaker-python-sdk/commit/4732545bd9d1880390b3951c9b941f2bb62d6e5a#r32336726

*Description of changes:*
Introduce a Dockerfile.ei to build TFS container with EI support.
Modify scripts to build and run Dockerfile.ei:
- build.sh: login to account to pull cpu image since Dockerfile.ei depends on Dockerfile.cpu; also add a build argument to pass in the AmazonEI model server binary
- shared.sh: add commands to parse the model variable

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://3r4crpxqvk.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=9d724aea-392d-4f2f-a37f-2a1a054c3f65%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: docker/Dockerfile.ei [hunk 1]
```diff
@@ -0,0 +1,10 @@
+ARG TFS_VERSION
+
+FROM 520713654638.dkr.ecr.us-west-2.amazonaws.com/sagemaker-tensorflow-serving:${TFS_VERSION}-cpu
+
+ARG TENSORFLOW_MODEL_SERVER
+
+# Validate the argument is specified
+RUN test $TENSORFLOW_MODEL_SERVER || exit 1
+
+COPY ./$TENSORFLOW_MODEL_SERVER /usr/bin/tensorflow_model_server
```
Comment (ChoiByungWook) [inline docker/Dockerfile.ei]: Is the ./ needed?

File: scripts/shared.sh [hunk 2]
```diff
@@ -32,6 +32,7 @@ function parse_std_args() {
     # defaults
     arch='cpu'
     version='1.12.0'
+    model='AmazonEI_TensorFlow_Serving_v1.12_v1'
```
Comment (ChoiByungWook) [inline scripts/shared.sh]: The file hosted in S3 is a zip, where does that get handled?

File: scripts/build.sh [hunk 3]
```diff
@@ -8,16 +8,28 @@ source scripts/shared.sh
...
     --build-arg TFS_SHORT_VERSION=$short_version \
+    --build-arg TENSORFLOW_MODEL_SERVER=$model \
     -f docker/Dockerfile.$arch \
     -t sagemaker-tensorflow-serving:$full_version-$arch \
     -t sagemaker-tensorflow-serving:$short_version-$arch container
+
+# logout 520713654638 account after building
+if [ "$arch" = "ei" ]; then
+    docker logout https://520713654638.dkr.ecr.$aws_region.amazonaws.com &>/dev/null
+fi
```
Comment (jesterhazy) [inline scripts/build.sh]: add nl @ eof

File: scripts/build.sh [hunk 4]
```diff
@@ -8,16 +8,28 @@ source scripts/shared.sh
...
 parse_std_args "$@"
 
-echo "pulling previous image for layer cache... "
-$(aws ecr get-login --no-include-email --registry-id $aws_account) &>/dev/null || echo 'warning: ecr login failed'
-docker pull $aws_account.dkr.ecr.$aws_region.amazonaws.com/sagemaker-tensorflow-serving:$full_version-$arch &>/dev/null || echo 'warning: pull failed'
-docker logout https://$aws_account.dkr.ecr.$aws_region.amazonaws.com &>/dev/null
+# Dockerfile.ei needs to login 520713654638 to pull the cpu image
+if [ "$arch" = "ei" ]; then
+    echo "pulling cpu image..."
+    $(aws ecr get-login --no-include-email --registry-id 520713654638) &>/dev/null || echo 'warning: ecr login failed'
```
Comment (jesterhazy) [inline scripts/build.sh]: no do not log in to prod repo. this breaks pr and release build flow.

File: docker/Dockerfile.ei [hunk 5]
```diff
@@ -0,0 +1,10 @@
+ARG TFS_VERSION
+
+FROM 520713654638.dkr.ecr.us-west-2.amazonaws.com/sagemaker-tensorflow-serving:${TFS_VERSION}-cpu
```
Comment (jesterhazy) [inline docker/Dockerfile.ei]: no do not use the prod repo! this breaks pr and release build flow. the image will be built from the previously released production image instead of the current image and we will ship a stale container!

File: scripts/shared.sh [hunk 6]
```diff
@@ -55,6 +56,11 @@ function parse_std_args() {
         shift
         shift
         ;;
+        -m|--model)
```
Comment (jesterhazy) [inline scripts/shared.sh]: `model` isn't a good name for something that points at a tfs executable

File: scripts/shared.sh (hunk 2 above)
Comment (jesterhazy) [reply scripts/shared.sh]: also 'model' isn't a good name

File: docker/Dockerfile.ei [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
+ARG TENSORFLOW_MODEL_SERVER
+
+# Validate the argument is specified
+RUN test $TENSORFLOW_MODEL_SERVER || exit 1
```
Comment (jesterhazy) [inline docker/Dockerfile.ei]: i don't think this check is needed. docker build should fail by itself it you don't specify, because the arg declaration has no default

File: scripts/build.sh [hunk 8]
```diff
@@ -8,16 +8,28 @@ source scripts/shared.sh
...
+    docker pull $aws_account.dkr.ecr.$aws_region.amazonaws.com/sagemaker-tensorflow-serving:$full_version-$arch &>/dev/null || echo 'warning: pull failed'
+    docker logout https://$aws_account.dkr.ecr.$aws_region.amazonaws.com &>/dev/null
+fi
 
 echo "building image... "
 docker build \
     --cache-from $aws_account.dkr.ecr.$aws_region.amazonaws.com/sagemaker-tensorflow-serving:$full_version-$arch \
     --build-arg TFS_VERSION=$full_version \
     --build-arg TFS_SHORT_VERSION=$short_version \
+    --build-arg TENSORFLOW_MODEL_SERVER=$model \
```
Comment (jesterhazy) [inline scripts/build.sh]: consider `EI_EXECUTABLE` or `TFS_EI_EXECUTABLE` instead

File: scripts/shared.sh (hunk 2 above)
Comment (jesterhazy) [reply scripts/shared.sh]: +1 the s3 fetch is missing, and there should be updates to the build.sh as well (or nothing is getting built).

also, if the S3 fetch normalized the file name when it downloaded, then there would likely be no need to pass the filename into the dockerfile as an ARG.

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://3r4crpxqvk.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=f66f48f4-9505-41b9-b7ab-aceff7eb17ba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by ChuyangDeng** [comment]
I've updated the scripts to handle zip files from s3 bucket, changed the EI dockerfile to avoid using prod account, dropped unnecessary arguments. I also updated README and add a functional test.

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://3r4crpxqvk.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=fdf9e4e9-0e83-41f4-8396-b9b56181efa0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: README.md [hunk 9]
```diff
@@ -67,6 +68,7 @@ in SageMaker, you need to publish it to an ECR repository in your account. The
 ```bash
 ./scripts/publish.sh --version 1.11 --arch cpu
 ./scripts/publish.sh --version 1.11 --arch gpu
+./scripts/publish.sh --version 1.11 --arch ei
```
Comment (jesterhazy) [inline README.md]: does 1.11 work for ei? if not we should update all three to current version

File: test/functional/test_elastic_inference.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+
+@pytest.fixture
+def pretrained_model_data(region):
+    return 's3://sagemaker-sample-data-{}/tensorflow/model/resnet/resnet_50_v2_fp32_NCHW.tar.gz'.format(region)
+
+
+@pytest.mark.skip_if_non_supported_ei_region
+@pytest.mark.skip_if_no_accelerator
+def test_deploy_elastic_inference_with_pretrained_model(pretrained_model_data, docker_image_uri, sagemaker_session, instance_type, accelerator_type):
```
Comment (jesterhazy) [inline test/functional/test_elastic_inference.py]: - endpoint needs to be removed at end of test
- not excited about taking a sagemaker sdk dependency for this. can we change this to use boto3 only? it shouldn't be too complicated since the model is already in s3

Comment (ChuyangDeng) [reply test/functional/test_elastic_inference.py]: ohhh, gotcha!

File: scripts/shared.sh [hunk 11]
```diff
@@ -28,6 +28,21 @@ function get_aws_account() {
     aws sts get-caller-identity --query 'Account' --output text
 }
 
+function get_tfs_executable() {
```
Comment (ChoiByungWook) [inline scripts/shared.sh]: Should this function utilize the parsed args below for the TF version?

It also looks like the naming scheme isn't consistent for the zip file for 1.11 and 1.12 already, which might require updating this file often.

File: test/functional/conftest.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (ChoiByungWook) [inline test/functional/conftest.py]: @jesterhazy, do you care about this being under functional or integration/sagemaker?

I think most of our framework containers are under integration/sagemaker, except for the sagemaker-tensorflow-container repository.

File: test/functional/test_elastic_inference.py (hunk 10 above)
Comment (ChoiByungWook) [reply test/functional/test_elastic_inference.py]: @ChuyangDeng 

https://github.com/aws/sagemaker-tensorflow-container/blob/66d62178248e112bd2cbe593aa3dcceec318a652/test/resources/python_sdk/timeout.py

File: test/functional/conftest.py (hunk 12 above)
Comment (jesterhazy) [reply test/functional/conftest.py]: yes, it seems a bit funky to have both. the stuff in test/integration are local docker container tests, so renaming test/functional to test/sagemaker or renaming everything to test/integration/{local, sagemaker} would fit the convention better

File: scripts/shared.sh (hunk 11 above)
Comment (ChuyangDeng) [reply scripts/shared.sh]: It only utilizes $version here, coz the naming for v1.11 and v1.12 are:
v1.11 -> Ubuntu -> Ubuntu.zip: Unbuntu/executable
v1.12 -> Ubuntu -> tfs_ei_v1_12_ubuntu.zip: v1_12_Ubuntu_2/executable

Even the v1.12's zip name and unzipped directory name are different. So I believe we will need to up date this file in the future.

File: test/functional/conftest.py (hunk 12 above)
Comment (ChuyangDeng) [reply test/functional/conftest.py]: I'll rename it to test/sagemaker

File: scripts/shared.sh (hunk 11 above)
Comment (ChoiByungWook) [reply scripts/shared.sh]: It seems like in S3 it is always going to follow the pattern of: s3://amazonei-tensorflow/Tensorflow\ Serving/{version}/Ubuntu.

We can do a discovery on the item for example with:
aws s3 ls s3://amazonei-tensorflow/Tensorflow\ Serving/v1.11/Ubuntu/ | awk '{print $4}'
which should return Ubuntu.zip

Afterwards, I think we can probably also discover the unzipped name as well. For example:
find . -name AmazonEI_TensorFlow_Serving_v${version}_v1* -exec mv {} container/ \;

There are probably better commands than the ones I used above in my examples.

File: test/sagemaker/conftest.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+logging.getLogger('boto').setLevel(logging.INFO)
+logging.getLogger('botocore').setLevel(logging.INFO)
+logging.getLogger('factory.py').setLevel(logging.INFO)
+logging.getLogger('auth.py').setLevel(logging.INFO)
+logging.getLogger('connectionpool.py').setLevel(logging.INFO)
+
+
+def pytest_addoption(parser):
+    parser.addoption('--aws-id')
+    parser.addoption('--docker-base-name', default='functional-tensorflow-serving')
```
Comment (ChoiByungWook) [inline test/sagemaker/conftest.py]: is functional meant to correspond to test/functional?

File: test/sagemaker/conftest.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+@pytest.fixture(scope='session')
+def processor(request):
+    return request.config.getoption('--processor')
+
+
+@pytest.fixture(scope='session')
+def tag(request, framework_version, processor):
+    provided_tag = request.config.getoption('--tag')
+    default_tag = '{}-{}-py2'.format(framework_version, processor)
```
Comment (ChoiByungWook) [inline test/sagemaker/conftest.py]: do we ever specify python within the tag for this specific container?

File: test/sagemaker/test_elastic_inference.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    logger.info('deploying model to endpoint: {}'.format(endpoint_name))
+
+    client.create_endpoint_config(EndpointConfigName=endpoint_config_name,
+                                  ProductionVariants=_production_variants(model_name, instance_type, accelerator_type))
+
+    client.create_endpoint(EndpointName=endpoint_name,
+                           EndpointConfigName=endpoint_config_name)
+
+    try:
+        client.get_waiter('endpoint_in_service').wait(EndpointName=endpoint_name)
```
Comment (ChoiByungWook) [inline test/sagemaker/test_elastic_inference.py]: You don't have to do this, however I can foresee us reusing this logic, maybe it would be better to refactor to another file?

File: README.md [hunk 16]
```diff
@@ -106,6 +108,15 @@ checkers using `tox`:
 tox
 ```
 
+To test Elastic Inference with Accelerator, you will need an AWS account, publish your built image to ECR repository and run the following command:
```
Comment (ChoiByungWook) [inline README.md]: can you provide an example of the command needed to run the test?

Comment (ChoiByungWook) [reply README.md]: "To test against Elastic Inference, you will..."

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://3r4crpxqvk.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=725b5eb6-e408-4647-a532-28f602dce2fd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: README.md (hunk 16 above)
Comment (ChuyangDeng) [reply README.md]: I'll update these as well as the above comments.

File: test/sagemaker/conftest.py (hunk 14 above)
Comment (jesterhazy) [reply test/sagemaker/conftest.py]: nope

File: test/sagemaker/test_elastic_inference.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+
+    input_data = {'instances': np.random.rand(1, 1, 3, 3).tolist()}
+
+    response = runtime_client.invoke_endpoint(EndpointName=endpoint_name,
+                                              ContentType='application/json',
+                                              Body=json.dumps(input_data))
+    result = json.loads(response['Body'].read().decode())
+    assert result['predictions'] is not None
+
+    client.delete_endpoint(EndpointName=endpoint_name)
```
Comment (jesterhazy) [inline test/sagemaker/test_elastic_inference.py]: delete_endpoint should clear endpoint_config too
also needs to be in a `finally` block so delete happens even if test fails

File: test/sagemaker/conftest.py (hunk 13 above)
Comment (ChuyangDeng) [reply test/sagemaker/conftest.py]: It's meant to be sagemaker-tensorflow-serving ... I just accidentally changed it to 'functional' for some reason I dont remember.

**Comment by jesterhazy** [comment]
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://3r4crpxqvk.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=b637d0a2-2359-4a12-9e51-dc54fcf1ba43%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: test/sagemaker/test_elastic_inference.py [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+def pretrained_model_data(region):
+    return 's3://sagemaker-sample-data-{}/tensorflow/model/resnet/resnet_50_v2_fp32_NCHW.tar.gz'.format(region)
+
+
+def _timestamp():
+    return time.strftime("%Y-%m-%d-%H-%M-%S")
+
+
+def _execution_role(session):
+    return session.resource('iam').Role('SageMakerRole').arn
```
Comment (ChoiByungWook) [inline test/sagemaker/test_elastic_inference.py]: should we consider adding an argparser for the role?

File: test/sagemaker/conftest.py [hunk 19]
```diff
@@ -0,0 +1,96 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (laurenyu) [inline test/sagemaker/conftest.py]: should this be under `test/integration/sagemaker` to match the other repos?

Comment (ChuyangDeng) [reply test/sagemaker/conftest.py]: > renaming test/functional to test/sagemaker or renaming everything to test/integration/{local, sagemaker} would fit the convention better

I went with the option that involves fewer changes. but I'm also okay to put everything under test/integration/... 😺

Comment (laurenyu) [reply test/sagemaker/conftest.py]: if you do move it, do it in a different PR haha
